lib/engine/step/g_18_co/takeover.rb

```diff
@@ -0,0 +1,265 @@
+# frozen_string_literal: true
+
+require_relative '../base'
+require_relative '../tokener'
+
+module Engine
+  module Step
+    module G18CO
+      class Takeover < Base
+        include Tokener
+        ACTIONS = %w[place_token pass].freeze
+
+        def actions(entity)
+          return [] unless entity == current_entity
+
+          ACTIONS
+        end
+
+        def round_state
+          {
+            pending_takeover: nil,
+          }
+        end
+
+        def active?
+          takeover_in_progress
+        end
+
+        def current_entity
+          pending_takeover[:destination_corp]
+        end
+
+        def taken_entity
+          pending_takeover[:source_corp]
+        end
+
+        def place_count
+          pending_takeover[:place_count]
+        end
+
+        def pending_takeover
+          @round.pending_takeover || {}
+        end
+
+        def description
+          "Replace up to #{place_count} Tokens"
+        end
+
+        def available_cities
+          taken_entity.tokens.map(&:city).compact
+        end
+
+        def available_city(_entity, city)
+          available_cities.include?(city)
+        end
+
+        def available_hex(_entity, hex)
+          available_cities.map(&:hex).include?(hex)
+        end
+
+        def pass_description
+          'Discard Remaining Tokens'
+        end
+
+        def process_pass(_action)
+          @game.log << "#{current_entity.name} passes replacing remaining #{taken_entity.name} tokens"
+
+          close_corporation(taken_entity)
+        end
+
+        def process_place_token(action)
+          entity = action.entity
+
+          unless available_hex(entity, action.city.hex)
+            @game.game_error("Cannot place a token on #{action.city.hex.name}")
+          end
+
+          old_token = taken_entity.tokens.find { |t| t.city == action.city }
+          new_token = entity.unplaced_tokens.last
+          old_token.swap!(new_token)
+          pending_takeover[:place_count] -= 1
+
+          @game.log << "#{current_entity.name} replaces #{taken_entity.name} token on #{action.city.hex.name}"
+
+          return if place_count.positive? &&
+            taken_entity.placed_tokens.any? &&
+            current_entity.unplaced_tokens.any?
+
+          close_corporation(taken_entity)
+        end
+
+        def takeover_in_progress
+          return true unless current_entity.nil?
+
+          @game.corporations.dup.find do |source|
+            source == takeover_corporation(source)
+          end
+        end
+
+        def takeover_corporation(source)
+          return unless source.floated?
+
+          president_share_count = source.owner.num_shares_of(source)
+          source.corporate_share_holders.find do |c_s_h|
```

Comment:
i think in general you shouldn't be executing stuff in a find either, can you instead find the corporation and then execute it?